src/rust/lib/graph-editor/src/component/node/port.rs

```diff
@@ -0,0 +1,468 @@
+
+        // This is a conditional rotation that allows the port to either point inwards
+        // or outwards.
+        let pi             = Var::from(PI);
+        let rotation_angle = is_inwards * pi;
+        let rotation_angle = Var::<Angle<Radians>>::from(rotation_angle);
+        let sculpted_shape = sculpted_shape.rotate(rotation_angle);
+
+        sculpted_shape.into()
+    }
+
+    ensogl::define_shape_system! {
+        (height:f32,width:f32,inner_radius:f32,is_inwards:f32) {
+            // NOTE: `is_inwards` should only be 0.0 or 1.0.
+            new_port(height,width,inner_radius,is_inwards)
+        }
+    }
+
+    impl Shape{
+        /// Set the shape parameters derived from the `Specification`.
+        pub fn update_from_spec(&self,spec:&Specification){
+            self.height.set(spec.height);
+            self.inner_radius.set(spec.inner_radius);
+            self.width.set(spec.width.value.to_radians());
+            match &spec.direction{
+                Direction::In  => self.is_inwards.set(1.0),
+                Direction::Out => self.is_inwards.set(0.0),
+            };
+        }
+    }
+}
+
+
+
+// ============
+// === Port ===
+// ============
+
+/// Shape view for input port.
+#[derive(Debug,Default,Clone,Copy)]
+pub struct InputPortView {}
+impl ShapeViewDefinition for InputPortView {
+    type Shape = shape::Shape;
+    fn new(shape:&Self::Shape, _scene:&Scene,_shape_registry:&ShapeRegistry) -> Self {
+        let spec  = Specification::default();
+        shape.update_from_spec(&spec);
+        shape.is_inwards.set(1.0);
+
+        let bbox = Vector2::new(2.0 * shape.height.get(), 2.0 * shape.height.get());
+        shape.sprite.size().set(bbox);
+
+        Self {}
+    }
+}
+
+/// Shape view for output port.
+#[derive(Debug,Default,Clone,Copy)]
+pub struct OutputPortView {}
+impl ShapeViewDefinition for OutputPortView {
+    type Shape = shape::Shape;
+    fn new(shape:&Self::Shape, _scene:&Scene,_shape_registry:&ShapeRegistry) -> Self {
+        let spec  = Specification::default();
+        shape.update_from_spec(&spec);
+        shape.is_inwards.set(0.0);
+
+        let bbox = Vector2::new(2.0 * shape.height.get(), 2.0 * shape.height.get());
+        shape.sprite.size().set(bbox);
+
+        Self {}
+    }
+}
+
+/// Helper trait that describes a `ShapeViewDefinition` with a port shape.
+pub trait PortShapeViewDefinition = ShapeViewDefinition<Shape=shape::Shape>;
+
+/// Port definition.
+///
+/// A port must always be instantiated as an `InputPort` or an `OutputPort`. This determines its
+/// actual shape.
+///
+/// Example
+/// -------
+/// ```
+/// use graph_editor::component::node::port::InputPort;
+/// use graph_editor::component::node::port::OutputPort;
+///
+/// let input_port  = InputPort::default();
+/// let output_port = OutputPort::default();
+/// ```
+#[derive(Debug,CloneRef)]
+#[allow(missing_docs)]
+pub struct Port<T:PortShapeViewDefinition> {
+    pub data : Rc<PortData<T>>
+}
```

Comment:
Plase read about `CloneRef` from `prelude` - this is one of the ost important traits in our codebase. Apply it in your PR whenever it could be applied (you can derive it)

Reply:
I did implement this for `Port`. From my understanding, the other structures (like `Registry`) should not need it as they have clearer ownerships. (One owned `Registry` per `Node`).